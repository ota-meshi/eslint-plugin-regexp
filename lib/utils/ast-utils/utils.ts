import type { Rule, SourceCode, AST, Scope } from "eslint"
import * as eslintUtils from "eslint-utils"
import type {
    ArrowFunctionExpression,
    CallExpression,
    Expression,
    FunctionDeclaration,
    FunctionExpression,
    Identifier,
    Literal,
    MemberExpression,
    Node,
    RegExpLiteral,
} from "estree"
import { parseStringLiteral, parseStringTokens } from "../string-literal-parser"
import { baseParseReplacements } from "../replacements-utils"

/**
 * Get a parent node
 * The AST node used by ESLint always has a `parent`, but since there is no `parent` on Types, use this function.
 */
export function getParent<E extends Node>(node: Node | null): E | null {
    if (!node) {
        return null
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- ignore
    return (node as any).parent
}

/**
 * Find the variable of a given name.
 */
export function findVariable(
    context: Rule.RuleContext,
    node: Identifier,
): Scope.Variable | null {
    return eslintUtils.findVariable(getScope(context, node), node)
}

type SimpleVariable = Scope.Variable & {
    defs: [
        Scope.Definition & { type: "Variable" } & { node: { id: Identifier } },
    ]
}

/**
 * Finds a variable of the form `{var,let,const} identifier ( = <init> )?`.
 *
 * The returned variable is also guaranteed to have exactly one definition.
 *
 * @param context
 * @param expression
 */
function findSimpleVariable(
    context: Rule.RuleContext,
    identifier: Identifier,
): SimpleVariable | null {
    const variable = findVariable(context, identifier)

    if (!variable || variable.defs.length !== 1) {
        // we want a variable with 1 definition
        return null
    }

    const def = variable.defs[0]
    if (def.type !== "Variable" || def.node.id.type !== "Identifier") {
        // we want a simple variable
        return null
    }

    return variable as SimpleVariable
}

/**
 * Get the value of a given node if it's a constant of string.
 */
export function getStringIfConstant(
    context: Rule.RuleContext,
    node: Node,
): string | null {
    // Supports `regexp.source` that eslint-utils#getStringIfConstant does not track.
    if (
        node.type === "BinaryExpression" ||
        node.type === "MemberExpression" ||
        node.type === "Identifier" ||
        node.type === "TemplateLiteral"
    ) {
        const evaluated = getStaticValue(context, node)
        return evaluated && String(evaluated.value)
    }
    return eslintUtils.getStringIfConstant(node, getScope(context, node))
}

type GetStaticValueResult =
    | { value: unknown }
    | { value: undefined; optional?: true }

/**
 * Get the value of a given node if it's a static value.
 */
export function getStaticValue(
    context: Rule.RuleContext,
    node: Node,
): GetStaticValueResult | null {
    // Supports `regexp.source` that eslint-utils#getStaticValue does not track.
    if (node.type === "BinaryExpression") {
        if (node.operator === "+") {
            const left = getStaticValue(context, node.left)
            if (left == null) {
                return null
            }
            const right = getStaticValue(context, node.right)
            if (right == null) {
                return null
            }
            return {
                value:
                    // eslint-disable-next-line @typescript-eslint/restrict-plus-operands, @typescript-eslint/no-explicit-any -- ignore
                    (left.value as any) + right.value,
            }
        }
    } else if (node.type === "MemberExpression") {
        const propName = getPropertyName(node, context)
        if (propName === "source") {
            const object = getStaticValue(context, node.object)
            if (object && object.value instanceof RegExp) {
                return { value: object.value.source }
            }
        }
    } else if (node.type === "TemplateLiteral") {
        const expressions: GetStaticValueResult[] = []

        for (const expr of node.expressions) {
            const exprValue = getStaticValue(context, expr)
            if (!exprValue) {
                return null
            }
            expressions.push(exprValue)
        }
        let value = node.quasis[0].value.cooked
        for (let i = 0; i < expressions.length; ++i) {
            value += String(expressions[i].value)
            value += node.quasis[i + 1].value.cooked!
        }
        return { value }
    } else if (node.type === "Identifier") {
        const deRef = dereferenceVariable(context, node)
        if (deRef !== node) {
            return getStaticValue(context, deRef)
        }
    }
    return eslintUtils.getStaticValue(node, getScope(context, node))
}

/**
 * Gets the scope for the current node
 */
export function getScope(
    context: Rule.RuleContext,
    currentNode: Node,
): Scope.Scope {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- ignore
    const scopeManager: Scope.ScopeManager = (context.getSourceCode() as any)
        .scopeManager

    // eslint-disable-next-line @typescript-eslint/no-explicit-any -- ignore
    let node: any = currentNode
    for (; node; node = node.parent || null) {
        const scope = scopeManager.acquire(node, false)

        if (scope) {
            if (scope.type === "function-expression-name") {
                return scope.childScopes[0]
            }
            return scope
        }
    }

    return scopeManager.scopes[0]
}

/**
 * Find function node
 */
export function findFunction(
    context: Rule.RuleContext,
    id: Identifier,
): FunctionDeclaration | FunctionExpression | ArrowFunctionExpression | null {
    let target = id

    const set = new Set<Identifier>()
    for (;;) {
        if (set.has(target)) {
            return null
        }
        set.add(target)
        const calleeVariable = findVariable(context, target)
        if (!calleeVariable) {
            return null
        }
        if (calleeVariable.defs.length === 1) {
            const def = calleeVariable.defs[0]
            if (def.node.type === "FunctionDeclaration") {
                return def.node
            }
            if (
                def.type === "Variable" &&
                def.parent.kind === "const" &&
                def.node.init
            ) {
                if (
                    def.node.init.type === "FunctionExpression" ||
                    def.node.init.type === "ArrowFunctionExpression"
                ) {
                    return def.node.init
                }
                if (def.node.init.type === "Identifier") {
                    target = def.node.init
                    continue
                }
            }
        }
        return null
    }
}

export type KnownMethodCall = CallExpression & {
    callee: MemberExpression & { object: Expression; property: Identifier }
    arguments: Expression[]
}
/**
 * Checks whether given node is expected method call
 */
export function isKnownMethodCall(
    node: CallExpression,
    methods: Record<string, number>,
): node is KnownMethodCall {
    const mem = node.callee
    if (
        mem.type !== "MemberExpression" ||
        mem.computed ||
        mem.property.type !== "Identifier"
    ) {
        return false
    }
    const argLength = methods[mem.property.name]
    if (node.arguments.length !== argLength) {
        return false
    }
    if (node.arguments.some((arg) => arg.type === "SpreadElement")) {
        return false
    }
    const object = mem.object
    if (object.type === "Super") {
        return false
    }

    return true
}

interface BaseElement {
    type: string
    range: [number, number]
}
export type ReplacementElement =
    | CharacterElement
    | DollarElement
    | ReferenceElement
export interface CharacterElement extends BaseElement {
    type: "CharacterElement"
    value: string
}
export interface DollarElement extends BaseElement {
    type: "DollarElement"
    kind: "$" | "&" | "`" | "'"
}
// $1, $<name>
export interface ReferenceElement extends BaseElement {
    type: "ReferenceElement"
    ref: number | string
    refText: string
}

/**
 * Parse replacements string
 */
export function parseReplacements(
    context: Rule.RuleContext,
    node: Literal,
): ReplacementElement[] {
    const stringLiteral = parseStringLiteral(context.getSourceCode().text, {
        start: node.range![0],
        end: node.range![1],
    })
    const tokens = stringLiteral.tokens.filter((t) => t.value)

    return baseParseReplacements(tokens, (start, end) => {
        return {
            range: [start.range[0], end.range[1]],
        }
    })
}

/**
 * Creates source range from the given offset range of the value of the given
 * string literal.
 *
 * @param sourceCode The ESLint source code instance.
 * @param node The string literal to report.
 * @returns
 */
export function getStringValueRange(
    sourceCode: SourceCode,
    node: Literal & { value: string },
    startOffset: number,
    endOffset: number,
): AST.Range | null {
    if (!node.range) {
        // no range information
        return null
    }
    if (node.value.length < endOffset) {
        return null
    }

    try {
        const raw = sourceCode.text.slice(node.range[0] + 1, node.range[1] - 1)
        let valueIndex = 0
        let start: number | null = null
        for (const t of parseStringTokens(raw)) {
            const endIndex = valueIndex + t.value.length

            // find start
            if (
                start == null &&
                valueIndex <= startOffset &&
                startOffset < endIndex
            ) {
                start = t.range[0]
            }

            // find end
            if (
                start != null &&
                valueIndex < endOffset &&
                endOffset <= endIndex
            ) {
                const end = t.range[1]
                const nodeStart = node.range[0] + 1
                return [nodeStart + start, nodeStart + end]
            }

            valueIndex = endIndex
        }
    } catch {
        // ignore
    }

    return null
}

/**
 * Check if the given expression node is regexp literal.
 */
export function isRegexpLiteral(node: Expression): node is RegExpLiteral {
    return node.type === "Literal" && "regex" in node
}

/**
 * Check if the given expression node is string literal.
 */
export function isStringLiteral(
    node: Expression,
): node is Literal & { value: string } {
    return node.type === "Literal" && typeof node.value === "string"
}

/**
 * Returns the string value of the property name accessed.
 *
 * This is guaranteed to return `null` for private properties.
 *
 * @param node
 * @returns
 */
export function getPropertyName(
    node: MemberExpression,
    context?: Rule.RuleContext,
): string | null {
    const prop = node.property
    if (prop.type === "PrivateIdentifier") {
        return null
    }

    if (!node.computed) {
        return (prop as Identifier).name
    }
    if (context) {
        return getStringIfConstant(context, prop)
    }
    if (isStringLiteral(prop)) {
        return prop.value
    }
    return null
}

/**
 * Converts an range into a source location.
 */
export function astRangeToLocation(
    sourceCode: SourceCode,
    range: AST.Range,
): AST.SourceLocation {
    return {
        start: sourceCode.getLocFromIndex(range[0]),
        end: sourceCode.getLocFromIndex(range[1]),
    }
}

/**
 * If the given expression is the identifier of an owned variable, then the
 * value of the variable will be returned.
 *
 * Owned means that the variable is readonly and only referenced by this
 * expression.
 *
 * In all other cases, the given expression will be returned as is.
 *
 * Note: This will recursively dereference owned variables. I.e. of the given
 * identifier resolves to a variable `a` that is assigned an owned variable `b`,
 * then this will return the value of `b`. Example:
 *
 * ```js
 * const c = 5;
 * const b = c;
 * const a = b;
 *
 * foo(a);
 * ```
 *
 * Dereferencing `a` in `foo(a)` will return `5`.
 */
export function dereferenceOwnedVariable(
    context: Rule.RuleContext,
    expression: Expression,
): Expression {
    if (expression.type === "Identifier") {
        const variable = findSimpleVariable(context, expression)

        if (!variable) {
            // we want a variable with 1 definition
            return expression
        }

        const def = variable.defs[0]

        const grandParent = getParent(def.parent)
        if (grandParent && grandParent.type === "ExportNamedDeclaration") {
            // exported variables are not owned because they can be referenced
            // by modules that import this module
            return expression
        }

        // we expect there two be exactly 2 references:
        //  1. for initializing the variable
        //  2. the reference given to this function
        if (variable.references.length !== 2) {
            return expression
        }

        const [initRef, thisRef] = variable.references
        if (
            !(
                initRef.init &&
                initRef.writeExpr &&
                initRef.writeExpr === def.node.init
            ) ||
            thisRef.identifier !== expression
        ) {
            return expression
        }

        return dereferenceOwnedVariable(context, def.node.init)
    }

    return expression
}

/**
 * If the given expression is the identifier of a variable, then the value of
 * the variable will be returned if that value can be statically known.
 *
 * This method assumes that the value of the variable is immutable. This is
 * important because it means that expression that resolve to primitives
 * (numbers, string, ...) behave as expected. However, if the value is mutable
 * (e.g. arrays and objects), then the object might be mutated. This is because
 * objects are passed by reference. So the reference can be statically known
 * (the value of the variable) but the value of the object cannot be statically
 * known. If the object is immutable (e.g. RegExp and symbols), then they behave
 * like primitives.
 */
export function dereferenceVariable(
    context: Rule.RuleContext,
    expression: Expression,
): Expression {
    if (expression.type === "Identifier") {
        const variable = findSimpleVariable(context, expression)

        if (variable) {
            const def = variable.defs[0]
            if (def.node.init) {
                if (def.parent.kind === "const") {
                    // const variables are always what they are initialized to
                    return dereferenceVariable(context, def.node.init)
                }

                // we might still be able to dereference var and let variables,
                // they just have to never re-assigned
                const refs = variable.references

                const inits = refs.filter((r) => r.init).length
                const reads = refs.filter((r) => r.isReadOnly()).length
                if (inits === 1 && reads + inits === refs.length) {
                    // there is only one init and all other references only read
                    return dereferenceVariable(context, def.node.init)
                }
            }
        }
    }

    return expression
}
