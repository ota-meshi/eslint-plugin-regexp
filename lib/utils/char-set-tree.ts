import type { CharSet } from "refa"
import type { ToCharSetElement } from "regexp-ast-analysis"
import { toCharSet } from "regexp-ast-analysis"
import type { Element } from "regexpp/ast"
import type { RegExpContext } from "."

/**
 * A class that expresses a pattern that simply consumes characters, as a tree.
 *
 * e.g
 *
 * 'abc' ->
 * {
 *   charSet: a,
 *   next: [
 *     {
 *       charSet: b,
 *       next: [
 *         {
 *           charSet: c,
 *           next: []
 *         }
 *       ]
 *     }
 *   ]
 * }
 *
 * 'ab(c|d)' ->
 * {
 *   charSet: a,
 *   next: [
 *     {
 *       charSet: b,
 *       next: [
 *         {
 *           charSet: c,
 *           next: []
 *         },
 *         {
 *           charSet: d,
 *           next: []
 *         }
 *       ]
 *     }
 *   ]
 * }
 *
 * 'a(b|c)d' ->
 * {
 *   charSet: a,
 *   next: [
 *     {
 *       charSet: b,
 *       next: [
 *         {
 *           charSet: d,
 *           next: []
 *         }
 *       ]
 *     },
 *     {
 *       charSet: c,
 *       next: [
 *         {
 *           charSet: d,
 *           next: []
 *         }
 *       ]
 *     },
 *   ]
 * }
 */
export class CharSetTree {
    private readonly element: ToCharSetElement

    public readonly charSet: CharSet

    private readonly baseNextList: CharSetTree[]

    private cacheNext: CharSetTree[] | null = null

    private range?: { min: number; max: number }

    private endOfPattern: boolean

    private readonly cacheIncludes = new Map<CharSetTree, boolean>()

    public constructor(data: {
        element: ToCharSetElement
        charSet: CharSet
        baseNextList?: CharSetTree[]
        range?: { min: number; max: number }
        endOfPattern?: boolean
    }) {
        this.element = data.element
        this.charSet = data.charSet
        this.baseNextList = data.baseNextList || []
        this.range = data.range
        this.endOfPattern = data.endOfPattern || false
    }

    public get next(): readonly CharSetTree[] {
        if (this.range) {
            if (this.cacheNext) {
                return this.cacheNext
            }
            if (this.range.min > 1 || this.range.max > 1) {
                const nextMin = Math.max(this.range.min - 1, 1)
                const tree = new CharSetTree({
                    element: this.element,
                    charSet: this.charSet,
                    baseNextList: this.baseNextList,
                    range: {
                        min: nextMin,
                        max: this.range.max - 1,
                    },
                    endOfPattern: this.endOfPattern,
                })
                const results: CharSetTree[] = [tree]
                if (nextMin === 1) {
                    results.push(...this.baseNextList)
                }
                return (this.cacheNext = results)
            }
        }
        return this.baseNextList
    }

    public get end(): boolean {
        if (this.range) {
            if (this.range.min > 1) {
                return false
            }
        }
        return this.endOfPattern
    }

    public get raw(): string {
        return this.element.raw
    }

    /**
     * Checks whether the given tree is included.
     */
    public includes(target: CharSetTree): boolean {
        const cache = this.cacheIncludes.get(target)
        if (cache != null) {
            return cache
        }
        const result = includesWithoutCache(this)
        this.cacheIncludes.set(target, result)
        return result

        /** Without cache */
        function includesWithoutCache(self: CharSetTree) {
            if (self.charSet.isDisjointWith(target.charSet)) {
                return self.next.some((n) => n.includes(target))
            }
            return self.startsWith(target)
        }
    }

    public startsWith(target: CharSetTree): boolean {
        if (this.charSet.isDisjointWith(target.charSet)) {
            return false
        }

        // match start
        if (target.end) {
            return true
        }
        for (const tree of this.next) {
            for (const t of target.next) {
                if (tree.startsWith(t)) {
                    return true
                }
            }
        }
        return false
    }

    public addNextInternal(next: CharSetTree): void {
        this.baseNextList.push(next)
    }

    public setRangeInternal(min: number, max: number): void {
        this.range = { min, max }
    }

    public markAsEndOfPatternInternal(): void {
        this.endOfPattern = true
    }
}

export class CharSetTreeRoot {
    public readonly elements: readonly CharSetTree[]

    private constructor(elements: CharSetTree[]) {
        this.elements = elements
    }

    /**
     * Get the CharSetTreeRoot from the given element. Returns null if it is not a simple pattern.
     */
    public static fromElement(
        element: Element,
        regexpContext: RegExpContext,
    ): CharSetTreeRoot | null {
        return CharSetTreeRoot.fromElements([element], regexpContext)
    }

    /**
     * Get the CharSetTreeRoot from the given elements. Returns null if it is not a simple pattern.
     */
    public static fromElements(
        baseElements: Element[],
        regexpContext: RegExpContext,
    ): CharSetTreeRoot | null {
        const result = elementsToList(baseElements)
        if (!result) {
            // has complex pattern
            return null
        }
        result.markAsEndOfPattern()
        return new CharSetTreeRoot(result.list)

        /** Elements to list */
        function elementsToList(
            elements: Element[],
        ): { list: CharSetTree[]; markAsEndOfPattern: () => void } | null {
            type State = {
                readonly addNext: (tree: CharSetTree) => void
                readonly markAsEndOfPattern: () => void
            }
            const list: CharSetTree[] = []
            let state: State = {
                addNext: (t) => list.push(t),
                markAsEndOfPattern: () => {
                    // noop
                },
            }
            for (const element of elements) {
                if (
                    element.type === "Character" ||
                    element.type === "CharacterClass" ||
                    element.type === "CharacterSet"
                ) {
                    const tree = toTree(element)
                    state.addNext(tree)
                    state = {
                        addNext: (t) => tree.addNextInternal(t),
                        markAsEndOfPattern: () =>
                            tree.markAsEndOfPatternInternal(),
                    }
                } else if (
                    element.type === "CapturingGroup" ||
                    element.type === "Group"
                ) {
                    const groupList: CharSetTree[] = []
                    for (const alt of element.alternatives) {
                        const altList = elementsToList(alt.elements)
                        if (!altList) {
                            // has complex pattern
                            return null
                        }
                        groupList.push(...altList.list)
                    }
                    groupList.forEach(state.addNext)
                    state = {
                        addNext: (t) =>
                            groupList.forEach((tree) =>
                                tree.addNextInternal(t),
                            ),
                        markAsEndOfPattern: () =>
                            groupList.forEach((tree) =>
                                tree.markAsEndOfPatternInternal(),
                            ),
                    }
                } else if (element.type === "Quantifier") {
                    if (element.max === 0) {
                        continue
                    }
                    if (element.max > 10) {
                        // consider has complex pattern
                        return null
                    }
                    const elementList = elementsToList([element.element])
                    if (!elementList) {
                        // has complex pattern
                        return null
                    }
                    for (const tree of elementList.list) {
                        state.addNext(tree)
                        tree.setRangeInternal(element.min, element.max)
                    }

                    const newState: State = {
                        addNext: (t) =>
                            elementList.list.forEach((tree) =>
                                tree.addNextInternal(t),
                            ),
                        markAsEndOfPattern: () =>
                            elementList.list.forEach((tree) =>
                                tree.markAsEndOfPatternInternal(),
                            ),
                    }

                    if (element.min === 0) {
                        const selfState = state
                        state = {
                            addNext(t) {
                                selfState.addNext(t)
                                newState.addNext(t)
                            },
                            markAsEndOfPattern() {
                                selfState.markAsEndOfPattern()
                                newState.markAsEndOfPattern()
                            },
                        }
                    } else {
                        state = newState
                    }
                } else {
                    // has complex pattern
                    return null
                }
            }

            return {
                list,
                markAsEndOfPattern: state.markAsEndOfPattern,
            }
        }

        /** Element to tree */
        function toTree(element: ToCharSetElement): CharSetTree {
            return new CharSetTree({
                element,
                charSet: toCharSet(element, regexpContext.flags),
            })
        }
    }

    /**
     * Checks whether the given tree is included.
     */
    public includes(target: CharSetTreeRoot): boolean {
        for (const tree of this.elements) {
            for (const t of target.elements) {
                if (tree.includes(t)) {
                    return true
                }
            }
        }
        return false
    }
}
