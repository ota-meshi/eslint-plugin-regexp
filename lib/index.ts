import type { RuleModule } from "./types"
import { rules as ruleList } from "./utils/rules"
import recommended from "./configs/recommended"
import all from "./configs/all"

const configs = {
    recommended,
    all,
}

const rules = ruleList.reduce((obj, r) => {
    obj[r.meta.docs.ruleName] = r
    return obj
}, {} as { [key: string]: RuleModule })

export = {
    configs,
    rules,
}
