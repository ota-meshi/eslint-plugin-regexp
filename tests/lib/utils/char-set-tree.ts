import chai from "chai"
import assert from "assert"
import type { RegExpContext } from "../../../lib/utils"
import { parseFlags } from "../../../lib/utils"
import { jestSnapshotPlugin } from "mocha-chai-jest-snapshot"
import type { CharSetTree } from "../../../lib/utils/char-set-tree"
import { CharSetTreeRoot } from "../../../lib/utils/char-set-tree"
import { parseRegExpLiteral } from "regexpp"

chai.use(jestSnapshotPlugin())

function toTree(regex: RegExp) {
    const ast = parseRegExpLiteral(regex)
    return CharSetTreeRoot.fromElements(ast.pattern.alternatives[0].elements, {
        flags: parseFlags(""),
    } as RegExpContext)
}

describe("CharSetTree", function () {
    const regexes = [
        /abc/,
        /abcd/,
        /a[bc]d/,
        /ab(c|d)/,
        /a(b|c)d/,
        /ab?cd{3,6}/,
        /ab?cd?/,
        /a(b|c)d?(e|f|[ghi])/,
        /a*/,
        /a+/,
        /a{0,10}/,
        /a{0,11}/,
    ]
    describe("matching strings", function () {
        for (const regex of regexes) {
            it(regex.toString(), function () {
                const tree = toTree(regex)
                chai.expect(tree && treeToStrings(tree)).toMatchSnapshot()
            })
        }
    })
    describe("tree", function () {
        for (const regex of regexes) {
            it(regex.toString(), function () {
                const tree = toTree(regex)
                chai.expect(tree && normalizeTree(tree)).toMatchSnapshot()
            })
        }
    })
    describe("`includes` should return the expected result.", function () {
        const tests = [
            { base: /abcd/, target: /abcd/, result: true },
            { base: /abcd/, target: /abc/, result: true },
            { base: /abcd/, target: /bcd/, result: true },
            { base: /abcd/, target: /bc/, result: true },
            { base: /abc/, target: /abcd/, result: false },
            { base: /bcd/, target: /abcd/, result: false },
            { base: /bc/, target: /abcd/, result: false },
            { base: /abcd/, target: /\w[a-z]/, result: true },
            { base: /a\w[a-z]d/, target: /abcd/, result: true },
            { base: /a\w[a-z]/, target: /abcd/, result: false },
        ]

        for (const { base, target, result } of tests) {
            it(`${base.toString()} includes ${target.toString()}`, function () {
                const treeBase = toTree(base)
                const treeTarget = toTree(target)
                assert.deepStrictEqual(
                    result,
                    treeBase && treeTarget && treeBase.includes(treeTarget),
                )
            })
        }
    })
})

function treeToStrings(root: CharSetTreeRoot): string[] {
    const result: string[] = []
    for (const n of root.elements) {
        result.push(...toStrings(n))
    }
    return result

    function toStrings(tree: CharSetTree): string[] {
        const chars: string[] = []
        if (tree.charSet.size > 3) {
            chars.push(tree.raw)
        } else {
            for (const c of tree.charSet.characters()) {
                chars.push(String.fromCodePoint(c))
            }
        }
        if (!tree.next.length) {
            return chars
        }
        const results: string[] = []
        if (tree.end) {
            results.push(...chars)
        }
        for (const n of tree.next) {
            for (const next of toStrings(n)) {
                results.push(...chars.map((c) => c + next))
            }
        }
        return results
    }
}

function normalizeTree(root: CharSetTreeRoot): any {
    return root.elements.map(normalize)

    function normalize(tree: CharSetTree): any {
        const chars: string[] = []
        if (tree.charSet.size > 1) {
            chars.push(tree.raw)
        } else {
            for (const c of tree.charSet.characters()) {
                chars.push(String.fromCodePoint(c))
            }
        }
        return {
            charSet: chars.length === 1 ? chars[0] : chars.join(),
            end: tree.end,
            next: tree.next.map(normalize),
        }
    }
}
