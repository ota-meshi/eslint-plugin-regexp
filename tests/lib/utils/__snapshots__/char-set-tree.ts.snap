// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharSetTree matching strings /a(b|c)d/ 1`] = `
Array [
  "abd",
  "acd",
]
`;

exports[`CharSetTree matching strings /a(b|c)d?(e|f|[ghi])/ 1`] = `
Array [
  "abde",
  "abdf",
  "abdg",
  "abdh",
  "abdi",
  "abe",
  "abf",
  "abg",
  "abh",
  "abi",
  "acde",
  "acdf",
  "acdg",
  "acdh",
  "acdi",
  "ace",
  "acf",
  "acg",
  "ach",
  "aci",
]
`;

exports[`CharSetTree matching strings /a*/ 1`] = `null`;

exports[`CharSetTree matching strings /a+/ 1`] = `null`;

exports[`CharSetTree matching strings /a[bc]d/ 1`] = `
Array [
  "abd",
  "acd",
]
`;

exports[`CharSetTree matching strings /a{0,10}/ 1`] = `
Array [
  "a",
  "aa",
  "aaa",
  "aaaa",
  "aaaaa",
  "aaaaaa",
  "aaaaaaa",
  "aaaaaaaa",
  "aaaaaaaaa",
  "aaaaaaaaaa",
]
`;

exports[`CharSetTree matching strings /a{0,11}/ 1`] = `null`;

exports[`CharSetTree matching strings /ab(c|d)/ 1`] = `
Array [
  "abc",
  "abd",
]
`;

exports[`CharSetTree matching strings /ab?cd?/ 1`] = `
Array [
  "abc",
  "abcd",
  "ac",
  "acd",
]
`;

exports[`CharSetTree matching strings /ab?cd{3,6}/ 1`] = `
Array [
  "abcddd",
  "abcdddd",
  "abcddddd",
  "abcdddddd",
  "acddd",
  "acdddd",
  "acddddd",
  "acdddddd",
]
`;

exports[`CharSetTree matching strings /abc/ 1`] = `
Array [
  "abc",
]
`;

exports[`CharSetTree matching strings /abcd/ 1`] = `
Array [
  "abcd",
]
`;

exports[`CharSetTree tree /a(b|c)d/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "b",
        "end": false,
        "next": Array [
          Object {
            "charSet": "d",
            "end": true,
            "next": Array [],
          },
        ],
      },
      Object {
        "charSet": "c",
        "end": false,
        "next": Array [
          Object {
            "charSet": "d",
            "end": true,
            "next": Array [],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /a(b|c)d?(e|f|[ghi])/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "b",
        "end": false,
        "next": Array [
          Object {
            "charSet": "d",
            "end": false,
            "next": Array [
              Object {
                "charSet": "e",
                "end": true,
                "next": Array [],
              },
              Object {
                "charSet": "f",
                "end": true,
                "next": Array [],
              },
              Object {
                "charSet": "[ghi]",
                "end": true,
                "next": Array [],
              },
            ],
          },
          Object {
            "charSet": "e",
            "end": true,
            "next": Array [],
          },
          Object {
            "charSet": "f",
            "end": true,
            "next": Array [],
          },
          Object {
            "charSet": "[ghi]",
            "end": true,
            "next": Array [],
          },
        ],
      },
      Object {
        "charSet": "c",
        "end": false,
        "next": Array [
          Object {
            "charSet": "d",
            "end": false,
            "next": Array [
              Object {
                "charSet": "e",
                "end": true,
                "next": Array [],
              },
              Object {
                "charSet": "f",
                "end": true,
                "next": Array [],
              },
              Object {
                "charSet": "[ghi]",
                "end": true,
                "next": Array [],
              },
            ],
          },
          Object {
            "charSet": "e",
            "end": true,
            "next": Array [],
          },
          Object {
            "charSet": "f",
            "end": true,
            "next": Array [],
          },
          Object {
            "charSet": "[ghi]",
            "end": true,
            "next": Array [],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /a*/ 1`] = `null`;

exports[`CharSetTree tree /a+/ 1`] = `null`;

exports[`CharSetTree tree /a[bc]d/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "[bc]",
        "end": false,
        "next": Array [
          Object {
            "charSet": "d",
            "end": true,
            "next": Array [],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /a{0,10}/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": true,
    "next": Array [
      Object {
        "charSet": "a",
        "end": true,
        "next": Array [
          Object {
            "charSet": "a",
            "end": true,
            "next": Array [
              Object {
                "charSet": "a",
                "end": true,
                "next": Array [
                  Object {
                    "charSet": "a",
                    "end": true,
                    "next": Array [
                      Object {
                        "charSet": "a",
                        "end": true,
                        "next": Array [
                          Object {
                            "charSet": "a",
                            "end": true,
                            "next": Array [
                              Object {
                                "charSet": "a",
                                "end": true,
                                "next": Array [
                                  Object {
                                    "charSet": "a",
                                    "end": true,
                                    "next": Array [
                                      Object {
                                        "charSet": "a",
                                        "end": true,
                                        "next": Array [],
                                      },
                                    ],
                                  },
                                ],
                              },
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /a{0,11}/ 1`] = `null`;

exports[`CharSetTree tree /ab(c|d)/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "b",
        "end": false,
        "next": Array [
          Object {
            "charSet": "c",
            "end": true,
            "next": Array [],
          },
          Object {
            "charSet": "d",
            "end": true,
            "next": Array [],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /ab?cd?/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "b",
        "end": false,
        "next": Array [
          Object {
            "charSet": "c",
            "end": true,
            "next": Array [
              Object {
                "charSet": "d",
                "end": true,
                "next": Array [],
              },
            ],
          },
        ],
      },
      Object {
        "charSet": "c",
        "end": true,
        "next": Array [
          Object {
            "charSet": "d",
            "end": true,
            "next": Array [],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /ab?cd{3,6}/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "b",
        "end": false,
        "next": Array [
          Object {
            "charSet": "c",
            "end": false,
            "next": Array [
              Object {
                "charSet": "d",
                "end": false,
                "next": Array [
                  Object {
                    "charSet": "d",
                    "end": false,
                    "next": Array [
                      Object {
                        "charSet": "d",
                        "end": true,
                        "next": Array [
                          Object {
                            "charSet": "d",
                            "end": true,
                            "next": Array [
                              Object {
                                "charSet": "d",
                                "end": true,
                                "next": Array [
                                  Object {
                                    "charSet": "d",
                                    "end": true,
                                    "next": Array [],
                                  },
                                ],
                              },
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
      Object {
        "charSet": "c",
        "end": false,
        "next": Array [
          Object {
            "charSet": "d",
            "end": false,
            "next": Array [
              Object {
                "charSet": "d",
                "end": false,
                "next": Array [
                  Object {
                    "charSet": "d",
                    "end": true,
                    "next": Array [
                      Object {
                        "charSet": "d",
                        "end": true,
                        "next": Array [
                          Object {
                            "charSet": "d",
                            "end": true,
                            "next": Array [
                              Object {
                                "charSet": "d",
                                "end": true,
                                "next": Array [],
                              },
                            ],
                          },
                        ],
                      },
                    ],
                  },
                ],
              },
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /abc/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "b",
        "end": false,
        "next": Array [
          Object {
            "charSet": "c",
            "end": true,
            "next": Array [],
          },
        ],
      },
    ],
  },
]
`;

exports[`CharSetTree tree /abcd/ 1`] = `
Array [
  Object {
    "charSet": "a",
    "end": false,
    "next": Array [
      Object {
        "charSet": "b",
        "end": false,
        "next": Array [
          Object {
            "charSet": "c",
            "end": false,
            "next": Array [
              Object {
                "charSet": "d",
                "end": true,
                "next": Array [],
              },
            ],
          },
        ],
      },
    ],
  },
]
`;
